doc/lexer_tutorial.md

```diff
@@ -0,0 +1,192 @@
+    Tab,
+    Linefeed,
+}
+```
+
+Everything else is a comment. There are no invalid lexes, so we'll define our own error type, a void enum:
+
+```rust
+pub enum LexicalError {
+    // Not possible
+}
+```
+
+Now for the lexer itself. We'll take a string slice as its input. For each token we process, we'll want to know the character value, and the byte offset in the string where it begins. We can do that by wrapping the `CharIndices` iterator, which yields tuples of `(usize, char)` representing exactly that information.
+
+```rust
+use std::str::CharIndices;
+
+pub struct Lexer<'input> {
+    chars: CharIndices<'input>,
+}
+
+impl<'input> Lexer<'input> {
+    pub fn new(input: &'input str) -> Self {
+        Lexer { chars: input.char_indices() }
+    }
+}
+```
+
+(The lifetime parameter `'input` indicates that the Lexer cannot outlive the string it's trying to parse.)
+
+Let's review our rules:
+
+- For a space character, we output `Tok::Space`.
+- For a tab character, we output `Tok::Tab`.
+- For a linefeed (newline) character, we output `Tok::Linefeed`.
+- We skip all other characters.
+- If we've reached the end of the string, we'll return `None` to signal EOF.
+
+Writing a lexer for a language with multi-character tokens can get very complicated, but this is so straightforward, we can translate it directly into code without thinking very hard. Here's our `Iterator` implementation:
+
+```rust
+impl<'input> Iterator for Lexer<'input> {
+    type Item = Spanned<Tok, usize, LexicalError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            return Some(Ok(match self.chars.next() {
+                Some((i, ' ')) => (i, Tok::Space, i+1),
+                Some((i, '\t')) => (i, Tok::Tab, i+1),
+                Some((i, '\n')) => (i, Tok::Linefeed, i+1),
+
+                None => return None, // End of file
+                _ => continue, // Comment; skip this character
+            }))
+        }
+    }
+}
+```
+
+That's it. That's all we need.
+
+## Updating the parser
+
+To use this with LALRPOP, we need to expose its API to the parser. It's pretty easy to do, but also somewhat magical, so pay close attention. At the bottom of the grammar file, we need an `extern` block:
+
+```rust
+extern {
+    // ...
+}
+```
+
+Now we tell LALRPOP about the `Location` and `Error` types, as if we're writing a trait:
+
+```rust
+extern {
+    type Location = usize;
+    type Error = lexer::LexicalError;
+    
+    // ...
+}
+```
+
+We expose the `Tok` type by kinda sorta redeclaring it:
+
+```rust
+extern {
+    type Location = usize;
+    type Error = lexer::LexicalError;
+
+    enum lexer::Tok {
+        // ...
+    }
+}
+```
+
+Now we have to declare each of our terminals. For each variant of `Tok`, we pick what name the parser will see, and write a pattern of the form `name => lexer::Tok::Variant`, similar to how action code works in grammar rules. The name can be an identifier, or a string literal. We'll use the latter.
```

Comment:
Nit: The actual form is a bit more flexible. For example, what comes after the `=>` can be a full-on pattern, not just a variant name, and the tokens need not just be `"foo"` but can also be other identifiers like `foo`. I'm not sure how much to go into here though -- I like that this tutorial starts in a simple place; I'd expect a follow-up to show more of the full capabilities. But it seems worth mentioning that this is a simplified variant, or perhaps listing in the follow-ups.